markers/env_markers/provider.py

```diff
@@ -152,13 +152,24 @@ def add_prov(prov):
     else:
         allowed_providers = potential_providers
 
+    need_prov_keys = False
+    for filter in filters:
+        if isinstance(filter, ProviderFilter):
+            for filt in filter.classes:
+                if hasattr(filt, 'type_name'):
+                    need_prov_keys = True
+                    break
+
     for provider in allowed_providers:
         argvalues.append([provider])
         # Use the provider key for idlist, helps with readable parametrized test output
         if selector == ONE:
-            idlist.append(provider.category)
-        elif selector == ONE_PER_CATEGORY:
-            idlist.append(provider.type)
+            if need_prov_keys:
+                idlist.append(provider.type_name)
+            else:
+                idlist.append(provider.category)
+        elif selector == ONE_PER_CATEGORY or selector == ONE_PER_TYPE:
```

Comment:
**Question**: Can we have ONE_PER_CATEGORY return us 'cloud', 'infra' in case we need both kind of providers?
ex. pytest.mark.provider([CloudInfraProvider], scope='module', selector=ONE_PER_CATEGORY)

Reply:
> InfraProvider, ONE_PER_CATEGORY -> virtualcenter, rhevm...
> ^ this is correct
That actually wasn't correct. What @rlbabyuk would be correct.